clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -3278,6 +3284,9 @@ void SemaHLSL::processExplicitBindingsOnDecl(VarDecl *VD) {
           << static_cast<int>(RT);
     }
   }
+
+  if (!HasBinding && VD->getType()->isHLSLResourceRecord())
+    SemaRef.Diag(VD->getLocation(), diag::warn_hlsl_implicit_binding);
```

Comment:
I would think ActOnFinishBuffer always executes when declaring a resource. Is there some case where the diagnostic is emitted here and ActOnFinishBuffer isn't called?

Reply:
`ActOnFinishBuffer` executes when finishing a `cbuffer {...}` or `tbuffer {...}` block. The word "buffer" in that function name specifically refers to those blocks, not the various resource types that happen to be buffers. So the logic here is needed for the warnings on things like `RWBuffer<T> x;`